src/umpire/strategy/DynamicPool.cpp

```diff
@@ -114,21 +115,56 @@ DynamicPool::findFreeBlock(std::size_t bytes) const
   return iter;
 }
 
+void* DynamicPool::allocateFromResource(std::size_t bytes)
+{
+  void* ptr{nullptr};
+  try {
+    ptr = m_allocator->allocate(bytes);
+  } catch (...) {
+    UMPIRE_LOG(Error,
+               "\n\tMemory exhausted at allocation resource. "
+               "Attempting to give blocks back.\n\t"
+               << getCurrentSize() << " Allocated to pool, "
+               << getFreeBlocks() << " Free Blocks, "
+               << getInUseBlocks() << " Used Blocks\n"
+      );
+    releaseFreeBlocks();
+    UMPIRE_LOG(Error,
+               "\n\tMemory exhausted at allocation resource.  "
+               "\n\tRetrying allocation operation: "
+               << getCurrentSize() << " Bytes still allocated to pool, "
+               << getFreeBlocks() << " Free Blocks, "
+               << getInUseBlocks() << " Used Blocks\n"
+      );
+    try {
+      ptr = m_allocator->allocate(bytes);
+      UMPIRE_LOG(Error,
+                 "\n\tMemory successfully recovered at resource.  Allocation succeeded\n"
+        );
+    }
+    catch (...) {
+      UMPIRE_LOG(Error,
+                 "\n\tUnable to allocate from resource even after giving back free blocks.\n"
+                 "\tThrowing to let application know we have no more memory: "
+                 << getCurrentSize() << " Bytes still allocated to pool\n"
+                 << getFreeBlocks() << " Partially Free Blocks, "
+                 << getInUseBlocks() << " Used Blocks\n"
+        );
+      throw std::bad_alloc{};
```

Comment:
I'm not sure about changing the exception value here.  The previous implementation was intending to perform a re-throw of the previously thrown exception (by calling throw with no arguments).

Reply:
I didn't know that throw without args threw the previous exception. Should I switch it back to that?

Reply:
Please switch it back - our exceptions contain additonal info that bad_alloc won't, e.g. CUDA error converted to its string description.

Reply:
Yes, please switch it back so that umpire exception information information is re-thrown.